Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -340,6 +340,56 @@ public final class HTTPServerPipelineHandler: ChannelDuplexHandler, RemovableCha
         }
     }
 
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        // We're being removed from the pipeline. We need to do a few things:
+        //
+        // 1. If we have buffered events, deliver them. While we shouldn't be
+        //     re-entrantly called, we want to ensure that so we take a local copy.
+        // 2. If we are quiescing, we swallowed a quiescing event from the user: replay it,
+        //     as the user has hopefully added a handler that will do something with this.
+        // 3. Otherwise, if we have a read pending, we need to release it.
+        //
+        // The basic theory here is that if there is anything we were going to do when we received
+        // either a request .end or a response .end, we do it now because there is no future for us.
+        // We also need to ensure we do not drop any data on the floor.
+        //
+        // At this stage we are no longer in the pipeline, so all further content should be
+        // blocked from reaching us. Thus we can avoid mutating our own internal state any
+        // longer.
+        let bufferedEvents = self.eventBuffer
+        for event in bufferedEvents {
+            switch event {
+            case .channelRead(let read):
+                context.fireChannelRead(read)
+            case .halfClose:
+                context.fireUserInboundEventTriggered(ChannelEvent.inputClosed)
+            case .error(let error):
+                context.fireErrorCaught(error)
+            }
+        }
+
+
+        if case .quiescingWaitingForRequestEnd = self.lifecycleState {
+            context.fireUserInboundEventTriggered(ChannelShouldQuiesceEvent())
+        } else if self.readPending {
```

Comment:
quick Q: why aren't we doing the `read` if we're waiting for the `.end`?

Reply:
If `readPending` isn't true, we haven't swallowed a call to `read`, so we don't need to fire one. Presumably another one is coming.

Reply:
@Lukasa I meant if we have `.quiescingWaitingForRequestEnd == self.lifecycleState` AND `self.readPending == true`

Reply:
Oh yeah, that's an artefact of my previous version. In that version if we were quiescing then I'd close the channel, rather than replay the event. Good catch, will fix.

Reply:
Ok, amended. Sadly that made the patch a lot larger, as it requires we keep a bit better track of whether we've actually _completed_ the quiescing to avoid double-notifying.